projects/hipblaslt/tensilelite/Tensile/Tests/unit/test_VerifyGRsCompleteBeforeLr1s.py

```diff
@@ -0,0 +1,386 @@
+        """
+        Simple failure case where the last GR is guaranteed AFTER the first LR1 starts.
+        """
+        assert self.num_vmfma == 8
+
+        optSchedule = {
+            "SYNC": [[0, 4, 4, self.num_vmfma-1]],
+            "LRA0": [[-1]],
+            "LRB0": [[-1]],
+            "GRA":  [[0,0,0,0]],
+            "GRB":  [[0,0,0,0]],
+            "LRA1": [[3,6,6,6]],
+            "LRB1": [[3,6,6,6]]
+        }
+
+        syncCode = [
+            SWaitCnt(dscnt=0, vlcnt=-1, vscnt=-1, comment="Wait for LR0s"),
+            SWaitCnt(dscnt=-1, vlcnt=4, vscnt=-1, comment="Wait for GRs"),
+            SBarrier(comment="For GRs"),
+            SWaitCnt(dscnt=0, vlcnt=-1, vscnt=-1, comment="Wait for LR1s"),
+        ]
+
+        sched = ScheduleInfo(1, self.num_vmfma, optSchedule, syncCode, None, None)
+        status, message = verify_grs_complete_before_lr1s(sched, {"kernel": self.kernel})
+        assert not status, f"Schedule should have failed (first LRA1 before last GRA), but passed. {message}"
+        assert message == "Code path 0: GRA at index 0 is not valid. It is guaranteed by the SWait @ idx=4 which is after the first corresponding LR1 @ idx=3. Order must be GR > SWait > SBarrier > LR1."
+
+    def test_less_than_1_full_iteration_to_complete(self):
+        """
+        Passing case where the GRs have and SWait and SBarrier on them before a full iteration has completed.
+        """
+        assert self.num_vmfma == 8
+
+        optSchedule = {
+            "SYNC": [[0, 1, 2, self.num_vmfma-1]],
+            "LRA0": [[-1]],
+            "LRB0": [[-1]],
+            "GRA":  [[3]],
+            "GRB":  [[3]],
+            "LRA1": [[6]],
+            "LRB1": [[6]]
+        }
+
+        syncCode = [
+            SWaitCnt(dscnt=0, vlcnt=-1, vscnt=-1, comment="Wait for LR0s"),
+            SWaitCnt(dscnt=-1, vlcnt=0, vscnt=-1, comment="Wait for GRs"),
+            SBarrier(comment="For GRs"),
+            SWaitCnt(dscnt=0, vlcnt=-1, vscnt=-1, comment="Wait for LR1s"),
+        ]
+        sched = ScheduleInfo(1, self.num_vmfma, optSchedule, syncCode, None, None)
+        status, message = verify_grs_complete_before_lr1s(sched, {"kernel": self.kernel})
+        assert status, f"Schedule should have passed validation but did not. {message}"
+
+    def test_some_grs_less_than_1_full_iteration_to_complete(self):
+        """
+        Passing case where some GRs have and SWait and SBarrier on them before a full iteration has completed.
+        """
+        assert self.num_vmfma == 8
+
+        optSchedule = {
+            "SYNC": [[0, 3, 3, self.num_vmfma-1]],
+            "LRA0": [[-1]],
+            "LRB0": [[-1]],
+            "GRA":  [[1,2,3,4,5]],
+            "GRB":  [[1,2,3,4,5]],
+            "LRA1": [[6]],
+            "LRB1": [[6]]
+        }
+
+        syncCode = [
+            SWaitCnt(dscnt=0, vlcnt=-1, vscnt=-1, comment="Wait for LR0s"),
+            SWaitCnt(dscnt=-1, vlcnt=2, vscnt=-1, comment="Wait for GRs"),
```

Comment:
```suggestion
            SWaitCnt(dscnt=-1, vlcnt=4, vscnt=-1, comment="Wait for GRs"),
```
should work, but 5 shouldn't? 

1) SYNC appears before GRA and GRB in the list. 
2) There are 2 GRAs and 2 GRBs strictly before 3 (at 1, 2)
3) 2 + 2 = 4

Reply:
Neither 3 nor 4 would work. There are actually only 2 GRs scheduled before 3.
```
"GRA":  [[1,2,3,4,5]],
"GRB":  [[1,2,3,4,5]],
```
The odd indices are not actual global reads, they're incrementing the counter. But this did uncover a bug, I was not correctly guarding for odd number of GRA indices when using DirectToLDS.

Reply:
Damn, I always forget those pointers in the sequence.